### PR TITLE
feat: Merge all contracts into one ABI

### PR DESCRIPTION
**Pull request by kevaundray**
# Related issue(s)

Resolves # 

# Description

This:

-  Merges all methods in a contract into one ABI.
- Adds a ContractType to contract methods so we can distinguish between private/public/unconstrained

## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

## Documentation needs
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

**Comment by kevaundray** [comment]
Note: We may want to add a test to see what happens when two contracts in different modules are compiled with the same name

File: crates/noirc_driver/src/contract.rs [hunk 1]
```diff
@@ -2,10 +2,46 @@ use std::collections::BTreeMap;
...
+
+#[derive(serde::Serialize, serde::Deserialize)]
 pub struct CompiledContract {
     /// The name of the contract.
     pub name: String,
     /// Each of the contract's functions are compiled into a separate `CompiledProgram`
     /// stored in this `BTreeMap`.
-    pub functions: BTreeMap<String, CompiledProgram>,
+    #[serde(rename = "methods")]
+    pub functions: BTreeMap<String, ContractMethod>,
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_driver/src/contract.rs:46]: agreed

**Comment by kevaundray** [comment]
> There's similar changes which would have to be made to the wasm crate here in order to be consistent.
> 
> My preference on this is that we remove support for compiling contracts in TS for now as we're going to be making too many breaking changes. @kobyhallx

Ideally we have this for offsite, so they can compile and make tests in typescript. Its not a hard condition and if it makes the lives of tooling folks hard we can find an alternative. 

I'm guessing the painpoint you are referring to is when we are developing on master since, once we cut a release the ts and native should be synced up?

File: crates/noirc_driver/src/lib.rs [hunk 2]
```diff
@@ -204,13 +205,21 @@ impl Driver {
...
-            Ok((function_name, program))
+            Ok((
+                function_name,
+                ContractFunction {
+                    // Note: currently we mark all of the contract methods
+                    // as secret as we do not support public functions.
+                    func_type: contract::ContractFunctionType::Secret,
+                    function: program,
+                },
+            ))
```
Comment (jfecher) [inline crates/noirc_driver/src/lib.rs:216]:
```rs
            let function = ContractFunction {
                // Note: currently we mark all of the contract methods
                // as secret as we do not support public functions.
                func_type: contract::ContractFunctionType::Secret,
                function: program,
            };
            Ok((function_name, function))

Comment (jfecher) [reply crates/noirc_driver/src/lib.rs:216]:
```suggestion
            let program = self.compile_no_check(options, *function)?;
            // Note: currently we mark all of the contract methods
            // as secret as we do not support public functions.
            let func_type = contract::ContractFunctionType::Secret;
            let contract_function = ContractFunction { function, func_type };
            Ok((function_name, contract_function))
```

**Review by jfecher — APPROVED** [review]

**Pull request merged**